### PR TITLE
Reth Book - Stages framework/draft

### PR DESCRIPTION
**Pull request by 0xKitsune**
I started a draft for the stages chapter in the book but wanted to stop before going too far to get some feedback on the direction. I would love to hear any thoughts on the overall tone/direction, as well as if you think it should be more code heavy/technical (I tried to keep it very high level).

Im happy to adjust towards whatever direction might be best for the book. Let me know what you think.

### CODE REVIEW

File: book/stages/README.md [hunk 1]
```diff
@@ -0,0 +1,80 @@
+# Stages
+
+The `stages` lib plays a central role in syncing the node, maintaining state, updating the database and more. The stages involved in the Reth pipeline are the `HeaderStage`, `BodyStage`, `SendersStage`, and `ExecutionStage`. Each of these stages are queued up and stored within the Reth pipeline.
```
Comment (rkrasiuk) [inline book/stages/README.md:3]: pls add a comment that this list is nonexhaustive, more will be added shortly

File: book/stages/README.md [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+Filename: crates/stages/src/pipeline.rs
+```rust
+pub struct Pipeline<DB: Database> {
+    stages: Vec<QueuedStage<DB>>,
+    max_block: Option<BlockNumber>,
+    events_sender: MaybeSender<PipelineEvent>,
+}
+```
+
+When the node is first started, a new `Pipeline` is intialized and all of the stages are added into `Pipline.stages`. Then, the `Pipeline::run` function is called, which starts the pipline, executing all of the stages continuously in an infinite loop. This process syncs the chain, keeping everything up to date with the chain tip.
```
Comment (rkrasiuk) [inline book/stages/README.md:14]:
```suggestion
When the node is first started, a new `Pipeline` is initialized and all of the stages are added into `Pipeline.stages`. Then, the `Pipeline::run` function is called, which starts the pipeline, executing all of the stages continuously in an infinite loop. This process syncs the chain, keeping everything up to date with the chain tip. 
```

File: book/stages/README.md [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    /// Unwind the stage.
+    async fn unwind(
+        &mut self,
+        db: &mut StageDB<'_, DB>,
+        input: UnwindInput,
+    ) -> Result<UnwindOutput, Box<dyn std::error::Error + Send + Sync>>;
+}
+```
+
+To get a better idea of what is happening at each part of the pipeline, lets walk through what is going on under the hood within the `execution()` function at each stage, starting with `HeadersStage`.
```
Comment (rkrasiuk) [inline book/stages/README.md:43]:
```suggestion
To get a better idea of what is happening at each part of the pipeline, lets walk through what is going on under the hood within the `execute()` function at each stage, starting with `HeadersStage`.
```

File: book/stages/README.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+
+## HeadersStage
+
+The `HeadersStage` is responsible for syncing the block headers, validating the header integrity and writing the headers to the database. When the `execute()` function is called, the head of the chain is updated to the most recent block height previously executed by the stage. At this point, the node status is also updated with the most recent block height, hash and total difficulty. These values are used during any new eth/65 handshakes. After updating the head, a stream is established with other peers in the network to sync the missing chain headers between the most recent state stored in the database and the chain tip. This stage relies on the stream to return the headers in descending order staring from the chain tip down to the latest block in the database.
+
+Each value produced from the stream is a `Vec<SealedHeader>`.
```
Comment (rkrasiuk) [inline book/stages/README.md:49]: stream yields `SealedHeader` (not vec)

**Review by rkrasiuk — COMMENTED** [review]
@0xKitsune off to a good start! left some comments

**Comment by 0xKitsune** [comment]
@rkrasiuk Thanks for all the comments, Ill get those fixed up later today.

**Comment by akirillo** [comment]
this is for later, but when we host the code docs on docs.rs, we should include links to the docs.rs pages for the `stages` crate, `pipeline.rs`, and each of the stages we mention

**Comment by akirillo** [comment]
I like this - focusing on a human-readable "what this does" + "how these things fit together" as opposed to a granular "how this works" at the level of individual functions which should be handled by code docs. I don't think you need to get more technical than this, it's great

As you continue on this, one thing that should definitely be included before we push is some rationale explaining why we are using the staged sync architecture, i.e. why this design is a good one. Currently the best written rationale on this that I've found is in [this Erigon blog article](https://erigon.substack.com/p/staged-sync-and-short-history-of), but it would be great to get more seasoned input from the rest of team.

Maybe we could also touch on the developer story of e.g. adding a stage to the pipeline? Just trying to focus on the book being how users interact with the components, but maybe this is obvious? Not sure, would love to hear your thoughts as well

**Comment by akirillo** [comment]
FWIW, when you have the time, try running a `cargo doc --open` for the repo and seeing what sorts of code docs have been written, to check that there's a reasonable boundary between them and the book. E.g., take a look at the docs generated for the `Pipeline` struct (`target/doc/reth_stages/struct.Pipeline.html`), they also get into some high-level explanation with a nice graph and I would love your thoughts on where that sort of content should live

Both can be subject to change, don't consider anything set in stone!

File: book/stages/README.md (hunk 1 above)
Comment (0xKitsune) [reply book/stages/README.md:3]: Sounds good, just added a note about this.

File: book/stages/README.md (hunk 2 above)
Comment (0xKitsune) [reply book/stages/README.md:14]: Thanks for catching these, I added a spellchecker to my IDE so I can catch these earlier in the future.

**Comment by 0xKitsune** [comment]
> As you continue on this, one thing that should definitely be included before we push is some rationale explaining why we are using the staged sync architecture, i.e. why this design is a good one. Currently the best written rationale on this that I've found is in [this Erigon blog article](https://erigon.substack.com/p/staged-sync-and-short-history-of), but it would be great to get more seasoned input from the rest of team.

Sounds good, I will keep this in mind and also check out the link. Feel free to send anything else over that you think may be helpful.

> FWIW, when you have the time, try running a cargo doc --open for the repo and seeing what sorts of code docs have been written, to check that there's a reasonable boundary between them and the book.

Great, I will check this out later tonight as well.

File: book/stages/README.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
+# Stages
+
+The `stages` lib plays a central role in syncing the node, maintaining state, updating the database and more. The stages involved in the Reth pipeline are the `HeaderStage`, `BodyStage`, `SendersStage`, and `ExecutionStage` (note that this list is non-exhaustive, and more pipeline stages will be added in the near future). Each of these stages are queued up and stored within the Reth pipeline.
+
+Filename: crates/stages/src/pipeline.rs
```
Comment (gakonst) [inline book/stages/README.md:5]: Let's link to the actual file and line number?

File: book/stages/README.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+        input: UnwindInput,
+    ) -> Result<UnwindOutput, Box<dyn std::error::Error + Send + Sync>>;
+}
+```
+
+To get a better idea of what is happening at each part of the pipeline, lets walk through what is going on under the hood within the `execute()` function at each stage, starting with `HeadersStage`.
+
+## HeadersStage
+
+The `HeadersStage` is responsible for syncing the block headers, validating the header integrity and writing the headers to the database. When the `execute()` function is called, the head of the chain is updated to the most recent block height previously executed by the stage. At this point, the node status is also updated with the most recent block height, hash and total difficulty. These values are used during any new eth/65 handshakes. After updating the head, a stream is established with other peers in the network to sync the missing chain headers between the most recent state stored in the database and the chain tip. This stage relies on the stream to return the headers in descending order staring from the chain tip down to the latest block in the database.
```
Comment (gakonst) [inline book/stages/README.md:48]: We should eventually add cross-links to everything once we have them, e.g. `eth/65 handshakes` shoudl link to the book section that talks about it

**Review by gakonst — COMMENTED** [review]
love the direction and in favor of quick-merging these PRs and doing fast-followons instead of having them live for too long.

Let's also add a note in the top-level README that the book is continuously rendered at https://paradigmxyz.github.io/reth/?

File: book/stages/README.md (hunk 5 above)
Comment (0xKitsune) [reply book/stages/README.md:5]: Sounds good, I will update this.

**Comment by 0xKitsune** [comment]
> love the direction and in favor of quick-merging these PRs and doing fast-followons instead of having them live for too long.
> 
> Let's also add a note in the top-level README that the book is continuously rendered at https://paradigmxyz.github.io/reth/?

Sounds good to me, I am working on updating the draft and pushing to get the chapter all wrapped up before the end of the weekend.

File: book/stages/README.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+```
+
+When the node is first started, a new `Pipeline` is initialized and all of the stages are added into `Pipeline.stages`. Then, the `Pipeline::run` function is called, which starts the pipeline, executing all of the stages continuously in an infinite loop. This process syncs the chain, keeping everything up to date with the chain tip. 
+
+Each stage within the pipeline implements the `Stage` trait which provides function interfaces to get the stage id, execute the stage and unwind the state if there was an issue during the stage execution. 
+
+
+Filename: crates/stages/src/stage.rs
+```rust
+#[async_trait]
```
[the remark this answers is not in the document]
Comment (akirillo) [reply book/stages/README.md]: ooh, i can look into how one might do this, ya

File: book/stages/README.md [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+pub struct Pipeline<DB: Database> {
+    stages: Vec<QueuedStage<DB>>,
+    max_block: Option<BlockNumber>,
+    events_sender: MaybeSender<PipelineEvent>,
+}
+```
+
+When the node is first started, a new `Pipeline` is initialized and all of the stages are added into `Pipeline.stages`. Then, the `Pipeline::run` function is called, which starts the pipeline, executing all of the stages continuously in an infinite loop. This process syncs the chain, keeping everything up to date with the chain tip. 
+
+Each stage within the pipeline implements the `Stage` trait which provides function interfaces to get the stage id, execute the stage and unwind the state if there was an issue during the stage execution.
```
Comment (akirillo) [inline book/stages/README.md:16]:
```suggestion
Each stage within the pipeline implements the `Stage` trait which provides function interfaces to get the stage id, execute the stage and unwind the stage if there was an issue during the stage execution. 
```
nit: do you mean to say "state" here? not all stages touch EVM "state" so this may be a slightly confusing term

File: book/stages/README.md [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+}
+```
+
+To get a better idea of what is happening at each part of the pipeline, lets walk through what is going on under the hood within the `execute()` function at each stage, starting with `HeadersStage`.
+
+<br>
+
+## HeadersStage
+
+The `HeadersStage` is responsible for syncing the block headers, validating the header integrity and writing the headers to the database. When the `execute()` function is called, the head of the chain is updated to the most recent block height previously executed by the stage. At this point, the node status is also updated with the most recent block height, hash and total difficulty. These values are used during any new eth/65 handshakes. After updating the head, a stream is established with other peers in the network to sync the missing chain headers between the most recent state stored in the database and the chain tip. This stage relies on the stream to return the headers in descending order staring from the chain tip down to the latest block in the database.
```
Comment (akirillo) [inline book/stages/README.md:50]: This might be a good place to talk about why the headers stage fetches headers in descending order (as opposed to ascending order, like the other stages), to avoid long-range attacks

File: book/stages/README.md (hunk 8 above)
Comment (0xKitsune) [reply book/stages/README.md:16]: What I meant by this is `unwind the changes to the database`. I can update the language to reflect this or reword it to whatever makes the most sense.

File: book/stages/README.md [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct SealedHeader {
+    /// Locked Header fields.
+    header: Header,
+    /// Locked Header hash.
+    hash: BlockHash,
+}
+```
+
+Each `SealedHeader` is then validated to ensure that it has the proper parent. Note that this is only a basic response validation, and a full header validation adhering to the consensus spec occurs in the `downloader`. Following the basic header validation, each header is then written to the database. If a header is not valid or the stream encounters any other error, the error is propagated up through the stage execution, the db changes are unwound and the stage is resumed from the most recent valid state.
```
Comment (akirillo) [inline book/stages/README.md:67]: validation according to consensus is going to stick out to the reader, and at a glance this might be read as "this stage does not do validation according to consensus". additionally, it's not super clear what the `downloader` is here.

i think it'd be helpful to mention, either here or in the paragraph above, that the stream that is established is an implementation of the `HeaderDownloader` trait (and, I suppose, pull in the definition for that trait), and double-down on the point that the implementation of that trait must validate the downloaded header according to the consensus specification.

so basically we construct a `SealedHeader` for the current latest header, then we request from our peers to download (in reverse order) all headers since the `SealedHeader`, and it is in the implementation of the downloader that consensus validation occurs

Comment (akirillo) [reply book/stages/README.md:67]: sorry, i know that was a super verbose comment, just wanted to get it down - please let me know if any part of it could be clarified

File: book/stages/README.md [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+
+This process continues until all of the headers have been downloaded and and written to the database. Finally, the total difficulty of the chain's head is updated and the function returns `Ok(ExecOutput { stage_progress: current_progress, reached_tip: true, done: true })`, signaling that the header sync has completed successfully. 
+
+<br>
+
+## BodyStage
+
+Once the `HeadersStage` completes successfully, the `BodyStage` will start execution. The body stage downloads block bodies for all of the new block headers that were stored locally in the database. The `BodyStage` first determines which block bodies to download by checking if the block body has an ommers hash and transaction root. 
+
+An ommers hash is the Keccak 256-bit hash of the ommers list portion of the block. If you are unfamiliar with ommers blocks, you can [click here to learn more](https://ethereum.org/en/glossary/#ommer). Note that while ommers blocks were important for new blocks created during Ethereum's proof of work chain, Ethereum's proof of stake chain selects exactly one block proposer at a time, causing ommers blocks not to be needed in post-merge Ethereum.
```
Comment (akirillo) [inline book/stages/README.md:77]: this is a great explanation to add, nice

File: book/stages/README.md [hunk 12]
```diff
@@ -0,0 +1,162 @@
...
+        self.s.to_big_endian(&mut sig[32..64]);
+        sig[64] = self.odd_y_parity as u8;
+
+        secp256k1::recover(&sig, hash.as_fixed_bytes()).ok()
+    }
+```
+
+In an [ECDSA (Elliptic Curve Digital Signature Algorithm) signature](), the "r", "s", and "v" values are three pieces of data that are used to mathematically verify the authenticity of a digital signature. ECDSA is a widely used algorithm for generating and verifying digital signatures, and it is often used in cryptocurrencies like Ethereum.
+
+The "r" is the x-coordinate of a point on the elliptic curve that is calculated as part of the signature process. The "s" is the s-value that is calculated during the signature process. It is derived from the private key and the message being signed. Lastly, the "v" is the "recovery value" that is used to recover the public key from the signature, which is derived from the signature and the message that was signed. Together, the "r", "s", and "v" values make up an ECDSA signature, and they are used to verify the authenticity of the signed transaction.
```
Comment (akirillo) [inline book/stages/README.md:124]: Is there a resource we could link to here that goes more deeply into depth on ECDSA signatures, if readers are interested? (doesn't need to be in this PR)

Comment (akirillo) [reply book/stages/README.md:124]: maybe in a further version of the docs we can allude to EIP-155 changes here? idk, maybe too low level, don't have much preference

File: book/stages/README.md [hunk 13]
```diff
@@ -0,0 +1,162 @@
...
+
+After all headers and their corresponding transactions have been executed, all of the resulting state changes are applied to the database, updating account balances, account bytecode and other state changes. After applying all of the execution state changes, if there was a block reward, it is applied to the validator's account. 
+
+At the end of the `execute()` function, a familiar value is returned, `Ok(ExecOutput { done: is_done, reached_tip: true, stage_progress: last_block })` signaling a successful completion of the `ExecutionStage`.
+
+<br>
+
+# Next Chapter
+
+Now that we have covered all of the stages that are currently included in the `Pipeline`, you know how the Reth client stays synced with the chain tip and updates the database with all of the new headers, bodies, senders and state changes. While this chapter provides an overview, on how the pipeline stages work, the following chapter will dive deeper into the database, the networking stack and other exciting corners of the Reth codebase. Feel free to check out any of the codebase mentioned in this chapter and when you are ready, the next chapter will dive into the `database`.
```
Comment (akirillo) [inline book/stages/README.md:157]:
```suggestion
Now that we have covered all of the stages that are currently included in the `Pipeline`, you know how the Reth client stays synced with the chain tip and updates the database with all of the new headers, bodies, senders and state changes. While this chapter provides an overview on how the pipeline stages work, the following chapters will dive deeper into the database, the networking stack and other exciting corners of the Reth codebase. Feel free to check out any parts of the codebase mentioned in this chapter, and when you are ready, the next chapter will dive into the `database`.
```
(small nits)

File: book/stages/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (akirillo) [reply book/stages/README.md]: this should do the trick, ya? https://rust-lang.github.io/mdBook/format/mdbook.html#including-portions-of-a-file (i'm thinking we can just add `ANCHOR`s to the source code)

File: book/stages/README.md [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+    /// Returns Ok(false) if the
+    fn validate(&self, header: &SealedHeader, parent: &SealedHeader) -> Result<(), DownloadError> {
+        validate_header_download(self.consensus(), header, parent)?;
+        Ok(())
+    }
+}
+```
+
+
+The `HeaderStage` relies on the downloader stream to return the headers in descending order staring from the chain tip down to the latest block in the database. While other stages in the `Pipeline` start from the most recent block in the database up to the chain tip, the `HeaderStage` works in reverse to avoid [long-range attacks](https://messari.io/report/long-range-attack). When a node downloads headers in ascending order, it will not know if it is being subjected to a long-range attack until it reaches the most recent blocks. To combat this, the `HeaderStage` starts by getting the chain tip from the Consensus Layer, verifies the tip, and then walks backwards by the parent hash. Each value yielded from the stream is a `SealedHeader`.
```
Comment (akirillo) [inline book/stages/README.md:94]:
```suggestion
The `HeaderStage` relies on the downloader stream to return the headers in descending order starting from the chain tip down to the latest block in the database. While other stages in the `Pipeline` start from the most recent block in the database up to the chain tip, the `HeaderStage` works in reverse to avoid [long-range attacks](https://messari.io/report/long-range-attack). When a node downloads headers in ascending order, it will not know if it is being subjected to a long-range attack until it reaches the most recent blocks. To combat this, the `HeaderStage` starts by getting the chain tip from the Consensus Layer, verifies the tip, and then walks backwards by the parent hash. Each value yielded from the stream is a `SealedHeader`. 
```
nit: typo

File: book/stages/README.md [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub struct SealedHeader {
+    /// Locked Header fields.
+    header: Header,
+    /// Locked Header hash.
+    hash: BlockHash,
+}
+```
+
+Each `SealedHeader` is then validated to ensure that it has the proper parent. Note that this is only a basic response validation, and the `HeadderDownloader` uses the `validate` method during the `stream`, so that each header is validated according to the consensus specification before the header is yielded from the stream. After this, each header is then written to the database. If a header is not valid or the stream encounters any other error, the error is propagated up through the stage execution, the changes to the database are unwound and the stage is resumed from the most recent valid state.
```
Comment (akirillo) [inline book/stages/README.md:109]:
```suggestion
Each `SealedHeader` is then validated to ensure that it has the proper parent. Note that this is only a basic response validation, and the `HeaderDownloader` uses the `validate` method during the `stream`, so that each header is validated according to the consensus specification before the header is yielded from the stream. After this, each header is then written to the database. If a header is not valid or the stream encounters any other error, the error is propagated up through the stage execution, the changes to the database are unwound and the stage is resumed from the most recent valid state.
```
nit: typo

**Review by gakonst — APPROVED** [review]
Great start :) Let's merge in the interest of not making this too big and let's modify as we go.

**Review by gakonst — CHANGES_REQUESTED** [review]
The import errors in CI are because Rust's mdbook tries to run the README's stuff as unit tests, so you want to fix them first before merging. Alternatively, use the `ignore` / `no_compile` tag in the rustdocs, but I would not recommend that as the moment we refactor something you'll encounter an issue.  You can add the Reth / secp package imports https://github.com/rust-lang/mdBook/blob/master/Cargo.toml#L18-L36

**Review by gakonst — COMMENTED** [review]
https://rust-lang.github.io/mdBook/format/mdbook.html?highlight=ANCHOR#including-portions-of-a-file brilliant

**Review by gakonst — APPROVED** [review]
shipping and we can iterate. sick work guys. love it.
